### ISSUE TITLE
Confusing new `IFileSystemProviderWithFileReadCapability`

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#159175]
Refs: https://github.com/microsoft/vscode/commit/cd477d6fa18b6c6cc811a13b048b8bcd0462b860

I do not understand this condition:

https://github.com/microsoft/vscode/blob/9529e11f6481ae53bba821b05e34549491b9415e/src/vs/platform/files/common/files.ts#L575

How does that make sure a method `readFile` exists?

The only check for this that is valid currently is `FileReadWrite`, the `Readonly` is just a detail of the file system and not a capability.

**Comment by bpasero** [comment]
This looks like an actual typo to me, not sure the consequences, as it could lead to errors when wrongly asserting, so maybe moving to August for investigation.

**Comment by joyceerhl** [comment]
TLDR I misunderstood how the declared capabilities relate to the actual declared methods and the meaning of `FileSystemProviderCapabilities.Readonly`. For context, this change came in through https://github.com/microsoft/vscode/pull/157582 and the intention of this change was to support the scenario where an `IFileSystemProvider` has declared `readFile` but not `writeFile`: 

https://github.com/microsoft/vscode/blob/9529e11f6481ae53bba821b05e34549491b9415e/src/vs/platform/files/common/files.ts#L570-L572

since `readFile` is optional in the interface and the `FileReadWrite` capability requires both `readFile` and `writeFile` to be implemented: https://github.com/microsoft/vscode/blob/9529e11f6481ae53bba821b05e34549491b9415e/src/vs/platform/files/common/files.ts#L513

I made this change because when implementing a `IFileSystemProvider` with `readFile` but not `writeFile`, this would result in `withReadProvider` throwing here: https://github.com/microsoft/vscode/pull/157582/files#diff-9eff79f6e04dd0d895fdaac933118e59be90aebf9145c269fa0f23f517deeeecR152

I see how it's possible that a `FileSystemProvider` which has `FileSystemProviderCapabilities.Readonly` check does not have a `readFile` method implemented. Additionally I see that we have other examples of how to handle the requirement of a readonly filesystes which implements `readWrite`, e.g.:

https://github.com/microsoft/vscode/blob/2e30bf3e9f9399eeb1c1ea4a6574f5d6a41b640b/src/vs/workbench/services/extensions/browser/webWorkerFileSystemProvider.ts#L14
https://github.com/microsoft/vscode/blob/2e30bf3e9f9399eeb1c1ea4a6574f5d6a41b640b/src/vs/workbench/services/extensions/browser/webWorkerFileSystemProvider.ts#L46-L48

I am reverting https://github.com/microsoft/vscode/commit/cd477d6fa18b6c6cc811a13b048b8bcd0462b860. Apologies for the incorrect code change and thank you for pointing this out.
